client.go

```diff
@@ -1402,6 +1413,10 @@ var clientConnPool sync.Pool
 
 func (c *HostClient) releaseConn(cc *clientConn) {
 	cc.lastUseTime = time.Now()
+	if c.ConnectionIsValid != nil && !c.ConnectionIsValid(cc.c){
+		c.closeConn(cc)
+		return
+	}
```

Comment:
But this is exactly a such place, just before obtaining the lock
```
	cc.lastUseTime = time.Now()
	if c.ConnectionIsValid != nil && !c.ConnectionIsValid(cc.c){
		c.closeConn(cc)
		return
	}
	c.connsLock.Lock()
	c.conns = append(c.conns, cc)
	c.connsLock.Unlock()
```

As for whether it should be invalidated post-or-prior to request, my thoughts were:
- Invalidation is costly
- Invalidation is incoherent, as it's (presumably) based on some TTL and so on, which never will be 100% precise when network is considered.
- Invalidation hit happens much less than valid connections. Invalidated connection supposed to be exceptional. 

So, since it's not really coherent and potentially costly, I preferred to put additional potential work after issuing a request, making request happen faster.

So, this was an idea of invalidating after. That saying, I don't really mind it the other way, if there is a good reason

Reply:
I'm just thinking about the case where you use a connection once in a while and have idle connections sitting in the conns list of a long time. If in this case you want to invalidate based on a TTL it will never work. You are probably thinking about connection validation in a high QPS environment?

Maybe an option is to have two validation functions, one for after the request and one for before. But on the other hand that also makes the API even bigger and more complex again.

Reply:
Well, in case of long-idling connections my understanding, that they are supposed to be removed basing on one of these two:

```	// Keep-alive connections are closed after this duration.
	//
	// By default connection duration is unlimited.
	MaxConnDuration time.Duration

	// Idle keep-alive connections are closed after this duration.
	//
	// By default idle connections are closed
	// after DefaultMaxIdleConnDuration.
	MaxIdleConnDuration time.Duration```


And yes, my use case is actually benchmark and I don't want to potential nano/microseconds on synchronisation

And yeah, it actually sounds like an application that is using the library should decide what's best for it. Before, after or both.

How about having some enum, which passed to invalidator function, which will indicate, whether it's pre or post-validation? So by default, an implementation might as well just ignore  this argument and just do in both cases, making it most-coherent. And if someone(me :) ) want optimizations - then can return immediately if it's a pre-validation 

This way, signature of Client and HostClient still have just one struct member added

Reply:
That's a very good idea yes. Can you add this to your pull request.